modules/dnn/src/dnn.cpp

```diff
@@ -2961,8 +2962,17 @@ AsyncMat Net::forwardAsync(const String& outputName)
     std::vector<LayerPin> pins(1, impl->getPinByAlias(layerName));
     impl->setUpNet(pins);
 
-    if (impl->preferableBackend != DNN_BACKEND_INFERENCE_ENGINE)
-        CV_Error(Error::StsNotImplemented, "Asynchronous forward for backend which is different from DNN_BACKEND_INFERENCE_ENGINE");
+    switch (impl->preferableBackend)
+    {
+    case DNN_BACKEND_INFERENCE_ENGINE:
+    case DNN_BACKEND_CUDA:
+        break;
+    default:
+        CV_Error(Error::StsNotImplemented, "Asynchronous forward for backend which is different from DNN_BACKEND_INFERENCE_ENGINE/DNN_BACKEND_CUDA");
+    }
+
+    if (impl->preferableBackend == DNN_BACKEND_CUDA)
+        return std::async(std::launch::async, [this, &outputName] { return forward(outputName); });
```

Comment:
Need to check for race conditions. The network stream is populated with operations on another thread. This allows the user to mess around with the network object when the thread spawned by `std::async` is accessing the network object.

The correct way might be to populate the stream on the caller's thread and then return an `std::future` which waits for synchronization.